daft/runners/dynamic_runner.py

```diff
@@ -0,0 +1,187 @@
+    Construction,
+    make_agg_instruction,
+    make_fanout_hash_instruction,
+    make_fanout_random_instruction,
+    make_filter_instruction,
+    make_map_partition_instruction,
+    make_merge_instruction,
+    make_project_instruction,
+)
+from daft.execution.dynamic_schedule import (
+    DynamicSchedule,
+    ScheduleCoalesce,
+    ScheduleFanoutReduce,
+    ScheduleFileRead,
+    ScheduleFileWrite,
+    ScheduleGlobalLimit,
+    ScheduleJoin,
+    ScheduleMaterialize,
+    SchedulePartitionRead,
+    SchedulePipelineInstruction,
+    ScheduleSort,
+)
+from daft.internal.rule_runner import FixedPointPolicy, Once, RuleBatch, RuleRunner
+from daft.logical import logical_plan
+from daft.logical.logical_plan import LogicalPlan, PartitionScheme
+from daft.logical.optimizer import (
+    DropProjections,
+    DropRepartition,
+    FoldProjections,
+    PruneColumns,
+    PushDownClausesIntoScan,
+    PushDownLimit,
+    PushDownPredicates,
+)
+from daft.runners.partitioning import PartitionCacheEntry
+from daft.runners.runner import Runner
+
+
+class DynamicRunner(Runner):
+    def __init__(self) -> None:
+        super().__init__()
+        # From PyRunner
+        self._optimizer = RuleRunner(
+            [
+                RuleBatch(
+                    "SinglePassPushDowns",
+                    Once,
+                    [
+                        DropRepartition(),
+                        PushDownPredicates(),
+                        PruneColumns(),
+                        FoldProjections(),
+                        PushDownClausesIntoScan(),
+                    ],
+                ),
+                RuleBatch(
+                    "PushDownLimitsAndRepartitions",
+                    FixedPointPolicy(3),
+                    [PushDownLimit(), DropRepartition(), DropProjections()],
+                ),
+            ]
+        )
+
+    def optimize(self, plan: LogicalPlan) -> LogicalPlan:
+        # From PyRunner
+        return self._optimizer.optimize(plan)
+
+    def run(self, plan: LogicalPlan) -> PartitionCacheEntry:
+        plan = self.optimize(plan)
+
+        schedule = schedule_logical_node(plan)
+        schedule = ScheduleMaterialize(schedule)
+
+        for next_construction in schedule:
+            if next_construction is None:
+                import IPython
+
+                IPython.embed()
+            assert next_construction is not None
+            self._build_partitions(next_construction)
+
+        final_result = schedule.result_partition_set()
+        pset_entry = self.put_partition_set_into_cache(final_result)
+        return pset_entry
+
+    def _build_partitions(self, partspec: Construction) -> None:
+        partitions = partspec.inputs
+
+        for instruction in partspec.instruction_stack:
+            partitions = instruction(partitions)
+
+        partspec.report_completed(partitions)
+
+
+def schedule_logical_node(node: LogicalPlan) -> DynamicSchedule:
```

Comment:
Could we make this a static method in `DynamicSchedule ` as `from_logical_node`? we could avoid all the imports here?

Reply:
how about making a factory class `DynamicScheduleFactory` with `from_logical_node` as a classmethod in another file?